Modal animations

maxokorokov opened the pull request:
Again this is inspired by #2817, not cherry-picked though, because there were several issues with the implementation.

This PR has 3 commits:
- fix for nested `transitionend` events in `ngbRunTransition` with tests
- disabling of animations for e2e tests
- modal animations implementation

### Modal animations

Key implementation points are:
- fade in animation starts on `modal.open()` and is reversible.
- fade out animation starts when modal is closed/dismissed and is not reversible.
- modal is focused at the same time as fade in animation starts
- focus is restored and all modal key/mouse event handling is disabled as soon as fade out animation starts
- had to refactor `modal-window.ts` quite a lot... :/
- a new API to notify when animation is finished

```ts
const modalRef = modal.open(content);

modalRef.opened.subscribe(() => { ... });
modalRef.closed.subscribe(() => { ... });
```

![Screen Shot 2020-05-27 at 23 23 19](https://user-images.githubusercontent.com/8074436/83073765-448c1080-a071-11ea-8b49-79dfd4c88f26.png)

maxokorokov commented:
After discussing with @fbasso, there will be at least 1 improvement here later:
- handling of `hide/show` classes should be done with `ngbRunTransition`, not with bindings

benouat commented:
> * handling of `hide/show` classes should be done with `ngbRunTransition`, not with bindings

Only here in Modal, or on every widget using animation you mean ?

benouat commented on src/util/transition/ngbTransition.spec.ts:
```diff
@@ -337,14 +337,75 @@ if (isBrowserVisible('ngbRunTransition')) {
     });
   });
+
+  describe('ngbRunTransition nesting', () => {
+
+    beforeEach(() => TestBed.configureTestingModule({declarations: [TestComponentNested]}));
+
+    it(`should ignore all inner element transitions`, (done) => {
+      const fixture = TestBed.createComponent(TestComponentNested);
+      fixture.detectChanges();
+
+      const outerEl = fixture.componentInstance.outer.nativeElement;
+      const innerEl = fixture.componentInstance.inner.nativeElement;
+
+      const nextSpy = createSpy();
+      const errorSpy = createSpy();
+
+      spyOn(fixture.componentInstance, 'onTransitionInnerEnd');
+      spyOn(fixture.componentInstance, 'onTransitionOuterEnd');
+
+      ngbRunTransition(outerEl, () => {
+        outerEl.classList.add('ngb-test-hide-outer');
+        innerEl.classList.add('ngb-test-hide-inner');
+      }, {animation: true, runningTransition: 'continue'}).subscribe(nextSpy, errorSpy, () => {
+        expect(fixture.componentInstance.onTransitionOuterEnd).toHaveBeenCalledTimes(2);
+        expect(fixture.componentInstance.onTransitionInnerEnd).toHaveBeenCalledTimes(1);
+        expectOpacity(outerEl, '0');
+        expectOpacity(innerEl, '0');
+        expect(nextSpy).toHaveBeenCalledWith(undefined);
+        expect(errorSpy).not.toHaveBeenCalled();
+        done();
+      });
+    });
+  });
```
Wow, we might need a comment here with a schema and explanations. Schema with DOM structure, what node are transitioned, in which order. Because here, I can not extract any sense out of this test like that. I don't understand at all what's going on.

**EDIT**: I know the DOM is below in the component definition. But still, wouldn't hurt to have a description colocated with the test.

maxokorokov replied:
Yeah, the DOM is there... I mean do you want me to copy DOM as a comment or what? 
In general I agree that these tests are not easy to read... 😕

benouat replied:
> I mean do you want me to copy DOM as a comment or what?

No!! 😰 but sometimes, test component should be colocated to the test, and not at the end of the file.

benouat commented on src/util/transition/ngbTransition.ts:
```diff
@@ -74,7 +74,8 @@ export const ngbRunTransition =
           // guarantees, that we'll release the DOM element and complete 'ngbRunTransition'.
-          const transitionEnd$ = fromEvent(element, 'transitionend').pipe(takeUntil(stop$));
+          const transitionEnd$ =
+              fromEvent(element, 'transitionend').pipe(takeUntil(stop$), filter(({target}) => target === element));
```
Same as my previous comment. We should comment here that filter is to discard all _inner_ transition.

Question: what if the _inner_ transition is **longer** than the outer one ? Generally speaking we should wait for the longer one ?

maxokorokov replied:
Will add a comment, thanks. Just forgot to do this.

> we should wait for the longer one ?

No, I don't think so. We have no idea what would be inside, so we shouldn't wait. If they really need to wait for some transition inside, that can just make modal transition longer

fbasso commented:
> > * handling of `hide/show` classes should be done with `ngbRunTransition`, not with bindings
> 
> Only here in Modal, or on every widget using animation you mean ?

Yes. The purpose is to have a clear (and unique) way to manage the show/hide classes. The rule is :
- The show/hide classes are managed directly with the dom
- The fade class is managed by the binding. If this class is not here, the show/hide classes will not launch an animation.

It seems clearer (to me at least ! :) ) to manage these classes this way, instead of sometimes with the binding, sometimes with the dom.

maxokorokov commented on src/modal/modal-ref.ts:
```diff
@@ -106,22 +126,32 @@ export class NgbModalRef {
-    }
+      this._windowCmptRef = <any>null;
+      this._contentRef = <any>null;
+      this._closed.next();
+      this._closed.complete();
```
I was thinking about this too, but came to the conclusion, that we don't care about backdrop, because it's purely cosmetic. There is no key/mouse handling there or anything. Also wanted to avoid overcomplicating things here, because we have to `zip` + backdrop can be optional, etc.

> Suppose user override the fading transition for the backdrop to be 4s

As for this, I don't see why anybody would do this in real life to be honest :) it will be a matter of milliseconds and users will care about window itself, so I can't see how the app will be non-usable from the JS point of view. Otherwise - sure, while backdrop is there - you can't click anywhere, but this has nothing to do with JS logic to me.

But I can change this of course!
@fbasso do you have any input on this?

fbasso replied:
Good remark done by @benouat, it's a question we need to wonder, but I'm more in favor of keeping things simple. If no customization is done, the users don't see the difference, and if an issue is raised later, it will still be time to fix it

benouat replied:
> that we don't care about backdrop, because it's purely cosmetic

Actually I think we should care about it and I disagree with that. In my example, as soon as `closed` is fired, the app might trigger some UI updates to display things, and if backdrop is still transitioning, these changes will occur below the backdrop, which is bad.

I am convinced that `closed` should only emit after backdrop is gone (in the case there is a backdrop obviously)

maxokorokov replied:
Ok, will do the change for consistency. I was 50/50 on this one.

maxokorokov commented on src/modal/modal-backdrop.ts:
```diff
@@ -1,12 +1,38 @@
+    this._zone.onStable.asObservable().pipe(take(1)).subscribe(() => {
+      ngbRunTransition(
+          this._el.nativeElement, ({classList}) => classList.add('show'),
+          {animation: this.animation, runningTransition: 'continue'});
```
I'll open an issue where I'm going to list these kind of things I guess

maxokorokov replied:
Done → #3747. It would be simpler to handle these things separately, as they're refactorings across all components

maxokorokov commented:
@benouat, @fbasso to sum up `modal-window.ts` changes, Modal closing became imperative and introduced a bunch of private methods:

Before:

```ts
onAfterViewInit() {
  // 1. setup keyboard / mouse handling
  // 2. focus inside modal
}

onDestroy() {
  // 1. restore focus
  // 2. _closed$.next() to cleanup keyboard / mouse handling
}
```

After:

```ts
onAfterViewInit() { this._show(); }

_show() {
  // 1. run transition
  
  this._enableEventHandling();
  this._setFocus();
}

hide() {
  // 1. run transition

  this._disableEventHandling();
  this._restoreFocus();
}
```

And now that I think of it... I should probably put back `onDestroy()` that will do the `this._disableEventHandling();` cleanup "just-in-case"

benouat commented:
> I should probably put back `onDestroy()` that will do the `this._disableEventHandling();` cleanup "just-in-case"

yes!

maxokorokov commented:
Updated:
- [x] put back `onDestroy` that cleans up event handling
- [x] added comment about `.filter()` in the `ngbRunTransition`
- [x] put the `TestNestedComponent` next to the test case
- [x] firing `closed` event when both window and backdrop transitions finish

Todo:
- [ ] IE tests are unstable as usual

jnizet commented:
I have a design question regarding the new `NgbModal.closed` observable: why does it emit void instead of emitting the result passed to `NgbActiveModal.close()` (or an object wrapping the result and indicating if the modal was dismissed or closed). This would allow deprecating the `result` promise (which strangely uses a promise, which makes it unusual, and harder to mock in tests), and would allow ignoring dismissals with code such as:

```
const modalRef = modalService.open();
modalRef.close.pipe(
  filter(outcome => outcome.type = 'closed') // ignore dismissals
).subscribe(result => doSomethingWithTheResult(result)
```

In my experience, 99.9% of the time, dismissals are ignored, so maybe it would even be more user-friandly to have two observables: `closed` (`Observable<any>`, emitting the result) and `dismissed` (`Observable<any>`, emitting the reason), so that we can do

```
const modalRef = modalService.open();
modalRef.close.subscribe(result => doSomethingWithTheResult(result)
```

And if someone really wants to react to close and dismiss the same way, he/she can always merge the two observables.

jnizet commented:
Hmm, Now that I think of it, that would only provide the result once the animation is done, which would thus delay the action on the result. So it's probably not a good idea. But still, maybe it would be a good oppotunity to revisit the result promise and make it one or two observables, emitting and completing as soon as the modal is closed/dismissed (and not at the end of the animation)

benouat commented:
I agree with all the mambo/jumbo we are doing with the promise. It's clearly not a good DX. We should change that.

> completing as soon as the modal is closed/dismissed (and not at the end of the animation)

I disagree with that one (the in-parenthesis part). We should wait the end of the animation, otherwise (as I described it in my review) it could lead to weird UI behaviour. Giving back the hand too early could cause the application below the backdrop to update too soon, hence strange visual artifacts.

maxokorokov commented:
@jnizet, @benouat thanks for comments

- notification about the `result` of modal closing and end of animation should be separate things IMO. We could rename animation events to `shown/hidden` for more clarity maybe.
- I don't see much value in distinguishing between `close` and `dismiss` situations. I mean you as a developer will call `m.close('result')`, otherwise you could use `beforeDismiss: () => {}` check.
- I'd like to replace the `result` promise with an single observable that will systematically check `beforeDismiss`

I'd like to have this API in the end I think:

```ts
const modalRef = modal.open(content, { beforeDismiss: () => {} });

// when opening animation finishes
modalRef.opened.subscribe(() => {});

// when closing animation finishes and no more DOM
modalRef.closed.subscribe(() => {});

// when result is available (before "closed" emits) → to replace the promise (breaking change)
modalRef.result.subscribe(result => {});
```

However at the moment this PR has no breaking changes, that was my initial goal. I'd prefer we change the "result" promise separately, maybe in the same release, but not in this PR.

maxokorokov commented:
If we want to distinguish between close/dismiss, the alternative for me would be (also this is not a breaking change):

```ts
modal.shown; → animation end
modal.hidden; → animation end
modal.closed; → result
modal.dismissed; → result
```

jnizet commented:
I like that last proposal. And I guess we should show in the documentation how to deal with the issue that @benouat talked about: if you only want to deal with the result once the closing animation is done, then you can use

    zip(modal.closed, modal.hidden).subscribe(([result]) => doSomethingWithResultOnceClosed(result));

maxokorokov commented:
Also @fbasso pointed out, an alternative for the `zip` in demos might be:

```html
<button (click)="closeModal('reason')">
```

```ts
function closeModal(reason) {
  // 1. do things depending on the 'reason' if necessary
  // do stuff

  // 2. be notified when animation is finished
  this.modalRef.hidden.subscribe(() => {});

  // 3. actually trigger modal closing
  this.modalRef.close();
}
```

fbasso commented:
My 2 cents about it:

I know that it depends on the coding preference, but on my side, I would have expect a much simpler api. As the developer manage the action inside the dialog, he can do what he wants directly, instead of passing a value to the close function to be retrieved on the subscribe. For example (simplified code here):

```typescript
function actionClose() {
  
  // Code run on the save button, synchronously
  ....
  modal.close();

  modal.hidden.subscribe(() => {
    // Code run asynchronously, after the transition end
    // hidden is completed after the transition end, so no need to unsubscribe
  });
}

<ng-template #content let-modal>
  ....
<div class="modal-footer">
  <button type="button" (click)="actionClose()">Save</button>
</div>

</ng-template>
```

As the developer is the owner of the dialog buttons, he can do things directly instead of managing action in the observers.

fbasso commented:
@maxokorokov, you posted an answer quicker than me 😄

jnizet commented:
@fbasso @maxokorokov that's fine when a component opens a modal using an ng-template, because there is only one component involved, which does everything. But I usually use components for modal, and the responsibility is then split betwen two components:
 - one opens a modal and does something with its result when it's closed
 - one displays the modal body and closes it, passing the result to the first one

This is particularly important when the modal is generic and reusable, such as a generic  confirmation (Yes/no) modal.

maxokorokov commented:
What is merged for now is:

```ts
modal.shown; → animation end
modal.hidden; → animation end
modal.closed; → result
modal.dismissed; → result
```
